### DISCUSSION TITLE
Q&A: Developing and Testing (IDE, Production, ...)

### DISCUSSION THREAD

**Discussion by hideosnes** [automadcms/discussions#9] [Q&A]
Hei there,

I'm new to developing with Automad and I was wondering about tips, tricks and best practices to get started. (I*ll also update this thread along the way.) I would love to hear about your insights! Feel free to dump your knowledge that might help out a newbie such as myself to get started.

Here's some initial questions:

What IDE are you using?
What's your pipeline to develop with Automad up until shipping?
What server specs are you recommending?

**Comment by mdahmenzpf** [comment]
Hi there, great question!

Let me break it down.

## Editors

### Neovim

I personally use Neovim with this [config](https://github.com/marcantondahmen/nvim-config). This config has support for the Automad template syntax out of the box. In case you want to use your own config, there is also a separate [tree-sitter-automad](https://github.com/automadcms/tree-sitter-automad) package.

### VS Code

If you prefer VS Code, just install the Automad package as usual.

## Local Development

### Docker

You only need PHP and a webserver running. That's it. There are many ways to do this. One way is using Docker. This [repository](https://github.com/automadcms/automad-docker) can be a good starting point for creating a Docker image.

### Server Switcher (macOS & brew)

You can also just install Nginx or Apache with PHP directly on your machine. The [server-switcher](https://github.com/marcantondahmen/server-switcher) script helps you starting, stopping, configuring and changing servers.

### XAMPP
Another way is a bit more oldschool. Just use XAMPP as described in this [video ](https://www.youtube.com/watch?v=ieU-l6DubVg).

## Deployment

Since Automad is fully file-based, it is extremely simple to deploy. An entire project can be easily version controlled &mdash; Automad, packages and the content all in one repository. The actual deployment is then depending on the use case. It can generally be deployed using simple GitHub action using ssh and rsync.

### Use case 1: You develop a site, your client provides the content

You will want to develop locally and get your code deployed when pushing to GitHub using actions. Your client works in production and creates content that is rsynced from the production server and pushed to a content branch using ssh/rsync as well in a scheduled GitHub action. The detailled setup depends on the details of the project.

### Use case 2: You develop a site and provide the content as well

This is way easier. Just use a GitHub action that rsyncs the project with your production server. Everything is development and created locally.

### Use case 3: Manual deployment

Automad only contains simple files. Therefore you can also simply deploy a site from a local computer using ssh/rsync or similar.

## Specs

Automad should run just fine on low-end servers or any kind of shared hosting. Take a look at the system requirements [here](https://automad.org/getting-started/system-requirements). Version 2 may differ in requirements since it needs PHP 8.2+. Just make sure that caching is enabled.

Hope this helps!
